drivers/sgp30/sgp30.c

```diff
@@ -0,0 +1,353 @@
+#ifdef MODULE_SGP30_STRICT
+    dev->ready = false;
+    dev->_timer.callback = _read_cb;
+    dev->_timer.arg = dev;
+#endif
+
+    /* delay while powering up */
+    ztimer_sleep(ZTIMER_USEC, SGP30_POWER_UP_TIME);
+
+    /* read future set */
+    uint16_t version;
+    sgp30_read_future_set(dev, &version);
+    if (version < SGP30_REQUIRED_FEATURE_SET) {
+        DEBUG("[sgp30]: wrong version number, %04x instead of %04x\n",
+              version, SGP30_REQUIRED_FEATURE_SET);
+        return -1;
+    }
+
+    /* read serial id */
+    uint8_t serial[SGP30_SERIAL_ID_LEN];
+    if (sgp30_read_serial_number(dev, serial, SGP30_SERIAL_ID_LEN)) {
+        DEBUG_PUTS("[sgp30]: could not read serial number");
+        return -1;
+    }
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        DEBUG_PUTS("[sgp30]: serial");
+        for (uint8_t i = 0; i < SGP30_SERIAL_ID_LEN; i++) {
+            DEBUG("%02x ", serial[i]);
+        }
+        DEBUG_PUTS("\n");
+    }
+
+    /* start air quality measurement */
+    if (sgp30_start_air_quality(dev)) {
+        DEBUG_PUTS("[sgp30]: could not start air quality measurements ");
+        return -1;
+    }
+
+    return 0;
+}
+
+int sgp30_reset(sgp30_t *dev)
+{
+    assert(dev);
+    int ret = _rx_tx_data(dev, SGP30_CMD_SOFT_RESET, NULL, 0, SGP30_DELAY_SOFT_RESET, false);
+#ifdef MODULE_SGP30_STRICT
+    if (ret == 0) {
+        ztimer_remove(ZTIMER_USEC, &dev->_timer);
+        dev->ready = false;
+    }
+#endif
+    return ret ? -EPROTO : 0;
+}
+
+int sgp30_read_serial_number(sgp30_t *dev, uint8_t *buf, size_t len)
+{
+    assert(dev && buf && (len == SGP30_SERIAL_ID_LEN));
+    uint8_t frame[9];
+    if (_rx_tx_data(dev, SGP30_CMD_READ_SERIAL, (uint8_t *)frame, sizeof(frame),
+                    SGP30_DELAY_READ_SERIAL, true)) {
+        DEBUG_PUTS("[sgp30]: fail read");
+        return -EPROTO;
+    }
+    if (_crc8(frame, 2) != frame[2] || _crc8(frame + 3, 2) != frame[5] ||
+        _crc8(frame + 6, 2) != frame[8]) {
+        DEBUG_PUTS("[sgp30]: wrong crc");
+        return -EBADMSG;
+    }
+    memcpy(buf, frame, 2);
+    memcpy(&buf[2], &frame[3], 2);
+    memcpy(&buf[4], &frame[6], 2);
+    return 0;
+}
+
+int sgp30_read_future_set(sgp30_t *dev, uint16_t *version)
+{
+    uint8_t frame[3];
+
+    if (_rx_tx_data(dev, SGP30_CMD_GET_FEATURE_SET_VERSION, frame, sizeof(frame),
+                    SGP30_DELAY_GET_FEATURE_SET_VERSION, true)) {
+        return -EPROTO;
+    }
+    if (_crc8(frame, 2) != frame[2]) {
+        DEBUG_PUTS("[sgp30]: wrong crc");
+        return -EBADMSG;
+    }
+    *version = (frame[0] << 8) + frame[1];
+    return 0;
+}
+
+int sgp30_set_absolute_humidity(sgp30_t *dev, uint32_t a_humidity)
+{
+    /* max value is (255g/m3+255/256g/m3), or 255999 mg/m3*/
+    if (a_humidity > 256000) {
+        return -1;
+    }
+
+    /* scale down to g/m^3 */
+    uint16_t humidity_scaled =
+        (uint16_t)(((uint64_t)a_humidity * 256 * 16777) >> 24);
+
+    uint8_t frame[3];
```

Comment:
here and below would benefit for a setter (`_set(uint8_t *buf, uint16_t val) { buf[0] = val >> 8; buf[1] = val & 0xFF; buf[3] = _crc8(...);}`.